src/rust/engine/fs/src/store.rs

```diff
@@ -279,13 +276,30 @@ impl Store {
         expanded_digests
       })
       .and_then(move |digests| {
+        if Store::upload_is_faster_than_checking_whether_to_upload(digests.clone()) {
```

Comment:
This clone is going to copy the contents of the HashMap, which shouldn't be necessary. If you just pass a reference to digests here, and have `upload_is_faster_than_checking_whether_to_upload` take a reference to a HashMap, I think that should Just Work without the copying :)

Reply:
Ahh I tried this initially but I did it wrong so it confused me. But it makes sense now! Thank you!